js_modules/dagster-ui/packages/ui-core/src/app/CloudComponentContext.tsx

```diff
@@ -0,0 +1,16 @@
+import React, {useContext} from 'react';
+
+import {AppTopNavRightOfLogo} from './AppTopNav/AppTopNavRightOfLogo.oss';
+
+export const CloudComponentContext = React.createContext<{
+  AppTopNavRightOfLogo: (() => React.ReactNode) | React.MemoExoticComponent<() => React.ReactNode>;
+}>({
+  AppTopNavRightOfLogo,
+});
+
+export const componentStub = (component: keyof React.ContextType<typeof CloudComponentContext>) => {
+  return () => {
+    const {[component]: Component} = useContext(CloudComponentContext);
+    return <Component />;
+  };
+};
```

Comment:
Nice! Hmm curious to get @hellendag's thoughts here too:

- I was imagining that the file replacement approach (eg: having `RightOfLogo.oss.tsx` and replacing it in cloud + at build time with `RightOfLogo.cloud.tsx`) would mean that a React context was no longer required.  I'm not sure there's value in breaking all the files out separately if we still keep `CloudComponentContext`, because the context system also helps make it clear what the injection points are?  I do still like being able to spot the components that vary between oss and cloud though...

- `AppTopNavRightOfLogo.tsx` above does not seem to be imported anywhere? 

- I think it may be necessary, but I wish the injection point here wasn't the entire top right nav, because the cloud version will need to import and render `AppTopNavLinks`.  If we make a nice generic way of doing the injection with low overhead, it feels like we should try as hard as we can to inject at the level of individual buttons / toolbars / etc. where the "oss" version can be an empty stub...  Any chance we could push the site down to a no-op `<ExtraLinks />` inside the AppTopNavLinks link renderer?

I guess for me there are sort of two challenging bits in our current implementation -- the fact that we inject using different contexts / approaches in different places, and the fact that the injected code heavily imports little bits from the OSS code. Maybe the latter just can't really be fixed and is out-of-scope...